File: src/main/java/com/brandonitaly/bedrockskins/client/BedrockModelManager.java
package com.brandonitaly.bedrockskins.client;

import com.brandonitaly.bedrockskins.bedrock.BedrockFile;
import com.brandonitaly.bedrockskins.pack.SkinPackLoader;
import com.brandonitaly.bedrockskins.pack.SkinId;
import com.google.gson.Gson;

import java.util.HashMap;
import java.util.Map;

public final class BedrockModelManager {
    private BedrockModelManager() {}

    private static final Map<SkinId, BedrockPlayerModel> bedrockModels = new HashMap<>();
    private static final Gson gson = new Gson();

    public static BedrockPlayerModel getModel(SkinId skinId) {
        if (skinId == null) return null;

        if (bedrockModels.containsKey(skinId)) {
            return bedrockModels.get(skinId);
        }

        var skin = SkinPackLoader.getLoadedSkin(skinId);
        if (skin == null) return null;

        try {
            SkinPackLoader.registerTextureFor(skinId);
        } catch (Exception e) {
            e.printStackTrace();
        }

        try {
            BedrockFile bedrockFile = gson.fromJson(skin.geometryData, BedrockFile.class);
            var geometryList = bedrockFile.getGeometries();
            if (geometryList != null && !geometryList.isEmpty()) {
                var geometry = geometryList.get(0);
                var model = BedrockPlayerModel.create(geometry, false);
                bedrockModels.put(skinId, model);
                return model;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return null;
    }

    public static void clearAllModels() {
        bedrockModels.clear();
    }
}

File: src/main/kotlin/com/brandonitaly/bedrockskins/client/BedrockModelManager.kt
package com.brandonitaly.bedrockskins.client

import com.brandonitaly.bedrockskins.bedrock.BedrockFile
import com.brandonitaly.bedrockskins.pack.SkinPackLoader
import com.google.gson.Gson
import java.util.UUID

object BedrockModelManager {
    private val bedrockModels = mutableMapOf<UUID, BedrockPlayerModel>()
    private val activeSkinKeys = mutableMapOf<UUID, String>()
    private val gson = Gson()

    fun getModel(uuid: UUID): BedrockPlayerModel? {
        val skinKey = SkinManager.getSkin(uuid.toString()) ?: return null
        
        if (activeSkinKeys[uuid] != skinKey) {
            bedrockModels.remove(uuid)
            activeSkinKeys[uuid] = skinKey
        }

        if (bedrockModels.containsKey(uuid)) {
            return bedrockModels[uuid]
        }

        val skin = SkinPackLoader.loadedSkins[skinKey] ?: return null
        // Ensure the texture is registered on-demand so we don't eagerly load all textures
        try {
            SkinPackLoader.registerTextureFor(skinKey)
        } catch (e: Exception) {
            // Non-fatal: texture registration failure shouldn't prevent model creation
            e.printStackTrace()
        }
        
        try {
            val bedrockFile = gson.fromJson(skin.geometryData, BedrockFile::class.java)
            val geometry = bedrockFile.geometries?.firstOrNull()
            if (geometry != null) {
                val model = BedrockPlayerModel.create(geometry, false)
                bedrockModels[uuid] = model
                return model
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        return null
    }
}
